pkg/storaged/resize.jsx

```diff
@@ -0,0 +1,441 @@
+    let fsys;
+    let crypto_overhead;
+    let vdo;
+    let stratis_bdev;
+    let orig_size;
+    let block;
+
+    if (lvol_or_part.iface == "org.freedesktop.UDisks2.LogicalVolume") {
+        orig_size = lvol_or_part.Size;
+        block = client.lvols_block[lvol_or_part.path];
+        if (!block)
+            return lvol_or_part.Resize(size, { });
+    } else {
+        orig_size = lvol_or_part.Size;
+        block = client.blocks[lvol_or_part.path];
+    }
+
+    const crypto = client.blocks_crypto[block.path];
+    if (crypto) {
+        const cleartext = client.blocks_cleartext[block.path];
+        if (!cleartext)
+            return;
+        fsys = client.blocks_fsys[cleartext.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(cleartext);
+        stratis_bdev = client.blocks_stratis_blockdev[cleartext.path];
+        if (crypto.MetadataSize !== undefined)
+            crypto_overhead = crypto.MetadataSize;
+        else
+            crypto_overhead = block.Size - cleartext.Size;
+    } else {
+        fsys = client.blocks_fsys[block.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(block);
+        stratis_bdev = client.blocks_stratis_blockdev[block.path];
+        crypto_overhead = 0;
+    }
+
+    function fsys_resize() {
+        if (fsys) {
+            // HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1934567
+            //
+            // block_fsys.MountedAt might be out of synch with reality
+            // here if resizing the crypto container accidentally
+            // triggered an unmount.  Thus, we check synchronously
+            // whether or not we should be doing a offline resize or
+            // not.
+            //
+            // Another option for us would be to just mount the
+            // filesystem back if that's what we expect, to undo the
+            // bug mentioned above. But let's be a bit more passive
+            // here and hope the bug gets fixed eventually.
+            return (is_mounted_synch(client.blocks[fsys.path])
+                    .then(is_mounted => {
+                        // When doing an offline resize, we need to first repair the filesystem.
+                        if (!is_mounted) {
+                            return (fsys.Repair({ })
+                                    .then(function () { return fsys.Resize(size - crypto_overhead, { }) }));
+                        } else {
+                            return fsys.Resize(size - crypto_overhead, { });
+                        }
+                    }));
+        } else if (vdo) {
+            if (size - crypto_overhead > vdo.physical_size)
+                return vdo.grow_physical();
+            else if (size - crypto_overhead < vdo.physical_size)
+                return Promise.reject(_("VDO backing devices can not be made smaller"));
+            else
+                return Promise.resolve();
+        } else if (stratis_bdev) {
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19385-20230926-092314-7463d2dd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)